src/main/java/com/google/cloud/tools/appengine/AppEngineDescriptor.java

```diff
@@ -0,0 +1,111 @@
+ * Utilities to obtain information from appengine-web.xml.
+ */
+public class AppEngineDescriptor {
+
+  private static final String APP_ENGINE_NAMESPACE = "http://appengine.google.com/ns/1.0";
+  private final Document document;
+  
+  // private to force use of parse method
+  private AppEngineDescriptor(Document document) {
+    this.document = document;
+  }
+
+  /**
+   * Parses an appengine-web.xml file.
+   * 
+   * @param in the contents of appengine-web.xml
+   * @return a fully parsed object that can be queried 
+   * @throws IOException if parsing fails for any reason including malformed XML
+   */
+  public static AppEngineDescriptor parse(InputStream in) throws IOException {
+    try {
+      DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
+      documentBuilderFactory.setNamespaceAware(true);
+      return new AppEngineDescriptor(documentBuilderFactory.newDocumentBuilder().parse(in));
+    } catch (SAXException | ParserConfigurationException exception) {
+      throw new IOException("Cannot parse appengine-web.xml", exception);
+    }
+  }
+
+  /**
+   * @return project ID from the &lt;application&gt; element of the appengine-web.xml or null
+   *         if it is missing
+   * @throws AppEngineException if parsing the value fails
+   */
+  public String getProjectId()  {
+    return getTopLevelValue(document, "appengine-web-app", "application");
+  }
+
+  /**
+   * @return project version from the &lt;version&gt; element of the appengine-web.xml or
+   *         null if it is missing
+   * @throws AppEngineException if parsing the value fails
+   */
+  public String getProjectVersion() {
+    return getTopLevelValue(document, "appengine-web-app", "version");
+  }
+
+  /**
+   * @return service ID from the &lt;service&gt; element of the appengine-web.xml, or
+   *         null if it is missing. Will also look at module ID.
+   */
+  public String getServiceId() {
+    String serviceId = getTopLevelValue(document, "appengine-web-app", "service");
+    if (serviceId != null) {
+      return serviceId;
+    }
+    return getTopLevelValue(document, "appengine-web-app", "module");
+  }
+
+  private static String getTopLevelValue(Document doc, String parentTagName, String childTagName)
+      throws AppEngineException {
+    try {
+      NodeList parentElements = doc.getElementsByTagNameNS(APP_ENGINE_NAMESPACE, parentTagName);
+      if (parentElements.getLength() > 0) {
+        Node parent = parentElements.item(0);
+        if (parent.hasChildNodes()) {
+          for (int i = 0; i < parent.getChildNodes().getLength(); ++i) {
+            Node child = parent.getChildNodes().item(i);
+            if (child.getNodeName().equals(childTagName)) {
+              return child.getTextContent();
+            }
+          }
+        }
+      }
+      return null;
+    } catch (DOMException ex) {
+      // this shouldn't happen barring a very funky DOM implementation
+      throw new AppEngineException(ex);
```

Comment:
Since this is a "can't happen" case in an already Nullable method, I decided to go with returning null here instead of throwing an exception.